metatensor-torch/src/labels.cpp

```diff
@@ -218,11 +176,43 @@ TorchLabels LabelsHolder::view(const TorchLabels& labels, std::vector<std::strin
     return torch::make_intrusive<LabelsHolder>(std::move(names), std::move(new_values), CreateView{});
 }
 
-LabelsHolder::LabelsHolder(metatensor::Labels labels):
-    names_(names_from_metatensor(labels)),
-    values_(values_from_metatensor(labels)),
-    labels_(std::move(labels))
-{}
+LabelsHolder::LabelsHolder(metatensor::Labels labels): labels_(std::move(labels)) {
+    // extract the names
+    for (const auto* name: this->labels_->names()) {
+        this->names_.emplace_back(name);
+    }
+
+    // check if the labels are already associated with a tensor
+    auto* user_data = this->labels_->user_data();
+    if (user_data != nullptr) {
+        // if we start using user_data for more than this exact case (storing
+        // tensors inside Labels), this code might fails and will need to start
+        // checking that `user_data` is actually a tensor.
+        this->values_ = *static_cast<torch::Tensor*>(user_data);
+    } else {
+        auto clone = this->labels_.value();
+        // otherwise create a new tensor which share memory with the Labels.
+        const auto* values = clone.as_mts_labels_t().values;
+        auto sizes = std::vector<int64_t>{
+            static_cast<int64_t>(clone.count()),
+            static_cast<int64_t>(clone.size()),
+        };
+
+        this->values_ = torch::from_blob(
+            // This should really be a `const int32_t*`, but we can not prevent
+            // writing to this tensor since torch does not support read-only tensor:
+            // https://github.com/pytorch/pytorch/issues/44027
+            const_cast<int32_t*>(values),
+            sizes,
+            [clone=std::move(clone)](void*) mutable {
+                // capture `clone` inside the torch::Tensor custom deleter to
+                // keep them alive as long as the tensor might be
+                auto _ = std::move(clone);
```

Comment:
Shouldn't we set `user_data` in the else case? The other constructors set `user_data` in the Rust `labels` object. I don't know how well the rest of the TorchScript API manages the case of a Rust labels object without a set `user_data`. It just looks to me as an inconsistency that might cause problems somewhere, but don't really know.

Reply:
no, we don't. Either the labels where created by rust without user_data, and then we create values_ sharing memory with the labels. Or the labels where created by torch, and then we want to keep the corresponding tensor around, in case it is on a different device.

Setting user_data here would make it so the tensor keeps a reference to the labels (inside the lambda) and the labels keeps a reference to the tensor (in user_data), creating a reference cycle and leaking the corresponding memory.